rfcs/framework-0063-notification-system.md

```diff
@@ -0,0 +1,422 @@
+    timestamp: datetime
+
+
+@dataclass
+class Recipient:
+    """Recipient datamodel."""
+
+    # Recipient information (e.g. expanded user/group)
+    data: dict
+
+    # The "rendering" template to be used for this recipient
+    template: str
+
+    # Backend delivery options/config
+    backends: list[Backend]
+
+
+@dataclass
+class Backend:
+    """Backend delivery datamodel."""
+
+    # Unique backend identifier ("email", "slack", "cern-notification")
+    id: str
+
+    # Options/parameters for the backend to deliver the notification
+    params: dict
+```
+
+### API
+
+A Notification manager/client will be used for sending notifications as a Celery task.
+
+```python
+class NotificationManager:
+
+    def __init__(self, backends):
+        self.backends = backends
+
+    def validate(self, notification):
+        ...
+
+    # Client
+    def broadcast(self, notification: Notification, eager=False):
+        """Broadcast a notification via a Celery task."""
+        self.validate(notification)
+        task = broadcast_notification.si(notification)
+        return task.apply() if eager else task.delay()
+
+    # Consumer
+    def handle(self, notification: Notification):
+        """Handle a notification."""
+        for recipient in notification.recipients:
+            for backend in recipient.backends:
+                dispatch_notification(backend, recipient.data, notification.data)
+```
+
+The `broadcast_notification` and `dispatch_notification` tasks will be handling retries:
+
+```python
+# tasks.py
+@shared_task
+def broadcast(notification: Notification):
+    """Broadcast notification to backends according to event policy."""
+    current_notifications_manager.broadcast(notification)
+
+@shared_task
+def notify(notification, backend_id):
+    """Set notification and notify specific backend.
+
+    Will pass the key for the specific backend to notify.
+    """
+    current_notifications_manager.notify(notification, backend_id)
+```
+
+The final usage looks like the following:
+
+```python
+from invenio_notifications import current_notifications
+
+notification = ...  # see "Building notifications"
+current_notifications.broadcast(notification)
+```
+
+### Backends
+
+A notification backend will receive all the information from the notification system through a notification (i.e. type of action/event, user who triggered,  recipients, community/record (based on action/event)) to construct a suitable notification. The backend itself should *not* have to perform any queries (ElasticSearch or database). Its only purpose should be to take the received information, create the payload for the notification and send it to the recipients.
+
+```python
+from abc import ABC, abstractmethod
+
+class Backend(ABC):
+
+    id = None
+    """Unique ID of the backend."""
+
+    @abstractmethod
+    def send(self, notification, recipient, params):
+        """Here each concrete implementation will be able to consume the notification message."""
```

Comment:
```suggestion
        """Consume the notification message."""
```

Reply:
These should be clear now in the RFC, the manager class has the following high-level methods:

- `broadcast(notification)`: called by the "client", i.e. the service layer that triggers the notification. It does some basic validation and sends a Celery task with a minimal notification payload
- `handle_broadcast(notification)`: called by the Celery task from `broadcast(...)` above. Responsible for:
  - expanding the notification context via "read" calls to the service layer
  - building the list of recipients
  - filtering the list based on recipient preferences
  - figuring out what backend each recipient wants
  - sending of a Celery task for the actual delivery for each recipient backend
- `handle_dispatch(backend, recipient, notification)`: called by the Celery task for `handle_broadcast(...)` above. Responsible for calling the backend to deliver the message to a recipient